src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -2,23 +2,220 @@
+            // going to store the buffer in this list, create a new one 2x the size,
+            // and continue reading in the sequence.
+            // The data is no longer going to be contiguous in memory, but that doesn't
+            // matter since even if we were using the resize-every-time approach, we'd
+            // still have to resize again at the end to make the array exactly the right size.
+
+            // Here's a visualization showing what first, buffers, and current
+            // might look like for a 200-length enumerable:
+            //
+            // first: [items 0-31]
+            //
+            // buffers:
+            // [0]: [items 32-63]
+            // [1]: [items 64-127]
+            //
+            // current: [items 128-199], [slots 200-255 empty]
+
+            // The up-front allocation for the list will not be that much since
+            // the backing store will be Array.Empty<T>() if you do not pass in a
+            // capacity. It only starts allocating arrays when we add the first item.
+
+            var buffers = new List<T[]>(); // list of previous buffers
+            var current = new T[first.Length]; // the current buffer we're reading the sequence into
+            int read = first.Length; // number of items we've read so far, updated every time we exhaust a buffer
+
+            while (true)
+            {
+                int index = 0; // index into the current buffer
+
+                // Continue reading the data into the current buffer
+                do
+                {
+                    current[index++] = en.Current;
+
+                    if (index == current.Length)
+                    {
+                        index = -1;
+                        break;
+                    }
+                }
+                while (en.MoveNext());
+
+                // If the index is -1, we called Current but not MoveNext, however
+                // the loop expects to be entered with MoveNext already having been
+                // called. So we need to call MoveNext ourselves in that case.
+                if (index >= 0 || !en.MoveNext())
+                {
+                    // We've finished enumerating the sequence.
+                    // Copy the data from the first buffer, then walk the list of buffers
+                    // and copy the data from those. Finally, copy the data from the
+                    // buffer we were just processing.
+                    
+                    int remainder = index >= 0 ? index : current.Length; // If we got here from !en.MoveNext() that means index == -1 and there was just enough space
+                    int finalLength = read + remainder;
+
+                    var result = new T[finalLength];
+
+                    // Use Array.Copy for everything, since this codepath is only taken for
+                    // buffers of length > 32
+                    Array.Copy(first, 0, result, 0, first.Length);
```

Comment:
Rather than special-case first, should we just put first into buffers at the very beginning and then treat it like any other array in the list?  Then you can delete this copy.  The one-extra slot in buffer's underlying array of arrays should not matter.